6712835: Server compiler fails with assertion (loop_count < K,"infinite loop in PhaseIterGVN::transform")

diff -r 4aebfff4f8a2 hotspot/src/share/vm/opto/ifnode.cpp
--- openjdk.orig/hotspot/src/share/vm/opto/ifnode.cpp	Mon Sep 15 11:38:34 2008 +0200
+++ openjdk/hotspot/src/share/vm/opto/ifnode.cpp	Fri Sep 26 11:56:32 2008 +0200
@@ -569,6 +569,11 @@
   int true_path = phi->is_diamond_phi();
   if( true_path == 0 ) return NULL;
 
+  // Make sure that iff and the control of the phi are different. This
+  // should really only happen for dead control flow since it requires
+  // an illegal cycle.
+  if (phi->in(0)->in(1)->in(0) == iff) return NULL;
+
   // phi->region->if_proj->ifnode->bool->cmp
   BoolNode *bol2 = phi->in(0)->in(1)->in(0)->in(1)->as_Bool();
   
@@ -595,6 +600,7 @@
   }
 
   Node* new_bol = (flip ? phase->transform( bol2->negate(phase) ) : bol2);
+  assert(new_bol != iff->in(1), "must make progress");
   iff->set_req(1, new_bol);
   // Intervening diamond probably goes dead
   phase->C->set_major_progress();
